coderd/inboxnotifications_test.go

```diff
@@ -0,0 +1,723 @@
+		listTarget         string
+		listReadStatus     string
+		listStartingBefore string
+	}{
+		{"nok - wrong targets", `Query param "targets" has invalid values`, "", "wrong_target", "", ""},
+		{"nok - wrong templates", `Query param "templates" has invalid values`, "wrong_template", "", "", ""},
+		{"nok - wrong read status", "starting_before query parameter should be any of 'all', 'read', 'unread'", "", "", "erroneous", ""},
+	}
+
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+
+			client, _, _ := coderdtest.NewWithAPI(t, &coderdtest.Options{})
+			firstUser := coderdtest.CreateFirstUser(t, client)
+			client, _ = coderdtest.CreateAnotherUser(t, client, firstUser.OrganizationID)
+
+			ctx, cancel := context.WithTimeout(context.Background(), testutil.WaitLong)
+			defer cancel()
+
+			resp, err := client.Request(ctx, http.MethodGet, "/api/v2/notifications/inbox/watch", nil,
+				codersdk.ListInboxNotificationsRequestToQueryParams(codersdk.ListInboxNotificationsRequest{
+					Targets:        tt.listTarget,
+					Templates:      tt.listTemplate,
+					ReadStatus:     tt.listReadStatus,
+					StartingBefore: tt.listStartingBefore,
+				})...)
+			require.NoError(t, err)
+			defer resp.Body.Close()
+
+			err = codersdk.ReadBodyAsError(resp)
+			require.ErrorContains(t, err, tt.expectedError)
+		})
+	}
+
+	t.Run("OK", func(t *testing.T) {
+		t.Parallel()
+
+		ctx := testutil.Context(t, testutil.WaitLong)
+		logger := testutil.Logger(t)
+
+		db, ps := dbtestutil.NewDB(t)
+
+		firstClient, _, _ := coderdtest.NewWithAPI(t, &coderdtest.Options{
+			Pubsub:   ps,
+			Database: db,
+		})
+		firstUser := coderdtest.CreateFirstUser(t, firstClient)
+		member, memberClient := coderdtest.CreateAnotherUser(t, firstClient, firstUser.OrganizationID, rbac.RoleTemplateAdmin())
+
+		u, err := member.URL.Parse("/api/v2/notifications/inbox/watch")
+		require.NoError(t, err)
+
+		// nolint: bodyclose
+		wsConn, resp, err := websocket.Dial(ctx, u.String(), &websocket.DialOptions{
+			HTTPHeader: http.Header{
+				"Coder-Session-Token": []string{member.SessionToken()},
+			},
+		})
+		if err != nil {
+			if resp.StatusCode != http.StatusSwitchingProtocols {
+				err = codersdk.ReadBodyAsError(resp)
+			}
+			require.NoError(t, err)
+		}
+		defer wsConn.Close(websocket.StatusNormalClosure, "done")
+
+		inboxHandler := dispatch.NewInboxHandler(logger, db, ps)
+		dispatchFunc, err := inboxHandler.Dispatcher(types.MessagePayload{
+			UserID:                 memberClient.ID.String(),
+			NotificationTemplateID: notifications.TemplateWorkspaceOutOfMemory.String(),
+		}, "notification title", "notification content", nil)
+		require.NoError(t, err)
+
+		dispatchFunc(ctx, uuid.New())
+
+		_, message, err := wsConn.Read(ctx)
+		require.NoError(t, err)
+
+		var notif codersdk.GetInboxNotificationResponse
+		err = json.Unmarshal(message, &notif)
+		require.NoError(t, err)
+
+		require.Equal(t, 1, notif.UnreadCount)
```

Comment:
Is there a chance that `notifs.UnreadCount` will accidentally return 2 unread notifications? I haven't analyzed the code too deeply, but we have dealt with similar flakiness around insights tests (expected 1, returned 2).